crates/iceberg/src/spec/table_metadata.rs

```diff
@@ -461,9 +461,57 @@ impl TableMetadata {
         file_io: &FileIO,
         metadata_location: impl AsRef<str>,
     ) -> Result<()> {
+        use std::io::Write as _;
+
+        use flate2::write::GzEncoder;
+
+        let json_data = serde_json::to_vec(self)?;
+
+        // Check if compression is enabled via table properties
+        let codec = self
+            .properties
+            .get(crate::spec::table_properties::TableProperties::PROPERTY_METADATA_COMPRESSION_CODEC)
+            .map(|s| s.as_str())
+            .unwrap_or(crate::spec::table_properties::TableProperties::PROPERTY_METADATA_COMPRESSION_CODEC_DEFAULT);
+
+        let (data_to_write, actual_location) = match codec {
+            "gzip" => {
+                let mut encoder = GzEncoder::new(Vec::new(), flate2::Compression::default());
+                encoder.write_all(&json_data).map_err(|e| {
+                    Error::new(
+                        ErrorKind::DataInvalid,
+                        "Failed to compress metadata with gzip",
+                    )
+                    .with_source(e)
+                })?;
+                let compressed_data = encoder.finish().map_err(|e| {
+                    Error::new(ErrorKind::DataInvalid, "Failed to finish gzip compression")
+                        .with_source(e)
+                })?;
+
+                // Modify filename to add .gz before .metadata.json
+                let location = metadata_location.as_ref();
+                let new_location = if location.ends_with(".metadata.json") {
+                    location.replace(".metadata.json", ".gz.metadata.json")
```

Comment:
😄  https://iceberg.apache.org/spec/#naming-for-gzip-compressed-metadata-json-files

i wonder if its a good idea to just default to `metadata.json.gz` here and not create more files with the `.gz.metadata.json` suffix